File: Template Method/Ruby/TwoFactorAuthentication.rb
class TwoFactorAuthentication
    def getData(username)
        raise NotImplementedError
    end

    def sendNotification(username)
        raise NotImplementedError
    end

    def match(username)
        raise NotImplementedError
    end

    def authenticate(username, password)
        getData(username)

        sendNotification(username)

        match(username)
    end
end

File: Template Method/Javascript/TwoFactorAuthentication.js
class TwoFactorAuthentication {
  getData(username) {
    throw new Error('Not implemented');
  }

  sendNotification(username) {
    throw new Error('Not implemented');
  }

  match(username) {
    throw new Error('Not implemented');
  }

  authenticate(username, password) {
    this.getData(username);

    this.sendNotification(username);

    this.match(username);
  }
}

export default TwoFactorAuthentication;